### DISCUSSION TITLE
Best approach for selectors

### DISCUSSION THREAD

**Discussion by clenemt** [pmndrs/zustand#387] [Q&A] [answered]
Imagine you have a store as such:

```js
// store.js
const useStore = create(set => ({
  bears: [{ id: 1, attitude: 'angry'}, { id: 2, attitude: 'playful'}]
}))
```

and a component using that store: 

```js
// component.js
function AngryBear() {
  const bears = useStore(state => state.bears.filter(bear => bear.attitude === 'angry'))
  return <h1>{bears.length} angry bears around here</h1>
}
```

What would be the best way to memoize this selector to avoid unneeded work?

* Should I just put the filter inside the component after selecting the `state.bears` with a `useMemo` ?
* Should I put the selector in the store and access it through `state.angryBears` ?
* Other solution ?

Thanks :)

**Comment by dai-shi** [comment, marked as answer]
https://github.com/pmndrs/zustand#memoizing-selectors

<img width="677" alt="image" src="https://user-images.githubusercontent.com/490574/117794506-4225c500-b288-11eb-9bc2-5cfa52fb86c8.png">

Defining selectors outside component will be most preferable. In the future version (v4), this style will be more performant than inline selectors. If selectors depend on props, `useCallback` would be recommended. That said, it's rather unopinionated, so either way is fine.

(In this context, I assume the selector function is lightweight.)

**Reply by clenemt** [reply]
Thanks for the quick answer.

I now get it that this selector should be very lightweight as it is called every time any part of the state changes.
If i understand right then doing the selector like I did, would actually be a mistake since it would re-render the component each time the state changes.

```js
// Bad
const bears = useStore(state => state.bears.filter(bear => bear.attitude === 'angry'))

// Better
const bears = useStore(state => state.bears);
const angryBears = useMemo(() => bears.filter(bear => bear.attitude === 'angry'), [bears]);
```

So in order to share that `bears.filter()` which could be a little more complex, the recommendation is to expose a hook with that `useMemo()` wrapped?

```js
export function useAngryBears() {
  const bears = useStore(state => state.bears);
  return useMemo(() => bears.filter(bear => bear.attitude === 'angry'), [bears]);
}
```

**Reply by dai-shi** [reply]
> If i understand right then doing the selector like I did, would actually be a mistake since it would re-render the component each time the state changes.

You get it correct. You could also use custom `equalityFn` to avoid re-renders.
I would prefer your "Better" pattern with useMemo.

> So in order to share that bears.filter() which could be a little more complex, the recommendation is to expose a hook with that useMemo() wrapped?

Now, this works, but useMemo() only memoizes it across renders, not across components. If you goal is to avoid re-renders, it's just fine. If you want to reduce computation, it will need some more technique.

**Reply by clenemt** [reply]
> Now, this works, but useMemo() only memoizes it across renders, not across components. If you goal is to avoid re-renders, it's just fine. If you want to reduce computation, it will need some more technique.

Agreed 👍 

One thing that bothers me a bit though is the boilerplate of always creating that selector function, then creating that wrapper with useMemo, then finally using it in the component.

This is now deviating from my original question, but what would be a good way to derived data and share it across component. Would you recommend using something like reselect, proxy-memoize or other tools with zustand?

Maybe it is just me, but i think it would be nice to have note in the readme about all this 👍 

Thanks !

**Reply by dai-shi** [reply]
> what would be a good way to derived data and share it across component.

reselect and proxy-memoize would work. But if I were to do it in a simple way with zustand, I'd de-normalize the store, putting the derived data in state, trying not to break it (there's a risk).
And, I'd actually use jotai for such use cases, with many/complicated derived data.

> it would be nice to have note in the readme about all this

zustand is less opinionated about usage patterns. I'd like to have it in something like recipes instead of readme. For now, we can use a wiki page for such a recipe.

**Reply by dsacramone** [reply]
@clenemt What pattern did you end up with?

**Reply by maxwaiyaki** [reply]
This answer seems to have been removed in the docs. @dai-shi is this still the recommended approach?

**Reply by dai-shi** [reply]
Thanks for notifying. We removed the section, because v4 doesn't require it any more.

So, the recommended way is both patterns now:

```js
// this is totally fine.
const Component = () => {
  const bears = useBearStore(state => state.bears);
  // ...
};

// this is also fine. (theoretically, this is a little better without relying on JS optimization. But, it's premature optimization and it would be unnecessary for 99.9% of cases.)
const selectBears = state => state.bears;
const Component = () => {
  const bears = useBearStore(selectBears);
  // ...
};
```

**Comment by dannobytes** [comment]
Did the documentation for #memoizing-selectors go missing or get deleted? It doesn't seem to exist anymore on the readme.
https://github.com/pmndrs/zustand#memoizing-selectors

**Reply by dai-shi** [reply]
This was the discussion when we were assuming of using `useMutableSource` in v4.
v4 actually uses `useSyncExternalStore`, which doesn't require memoizing selectors, and the section is removed entirely. 
https://github.com/pmndrs/zustand/discussions/387#discussioncomment-3810458

**Reply by timeturnback** [reply]
@dai-shi show how exactly we should handle the case above ? 
```tsx
// JUST USE THIS
function AngryBear() {
  const bears = useStore(state => state.bears.filter(bear => bear.attitude === 'angry'))
  return <h1>{bears.length} angry bears around here</h1>
}

// SOLUTION 2 
const angrySelector = state => state.bears.filter(bear => bear.attitude === 'angry');
function AngryBear() {
  const bears = useStore(angrySelector)
  return <h1>{bears.length} angry bears around here</h1>
}
```
Or some way else

The problems is , for both solution . the `.filter()` will run every time the state change . may trigger heavy calculation . But if we use reselect . It will only trigger calculation when state.bears update . 

```ts
// reselect
import { createSelector } from 'reselect';

export const angrySelector = createSelector(
  (state) => state.bears,
  (bears, bearsList) => {
    console.log('totalAngryBear calc'); //
    return state.bears.filter(bear => bear.attitude === 'angry');
  }
);

function AngryBear() {
  const bears = useStore(angrySelector)
  return <h1>{bears.length} angry bears around here</h1>
}
```

My question is . Is this the best practice to handle state-calculation case like this

**Reply by dai-shi** [reply]
That's a different question from the OP. (It's about whether memoization is necessary for avoiding extra re-renders.)

If you want to avoid extra _calculation_, memoization is necessary.
reselect is one and proxy-memoize is another.
zustand is unopinionated about choosing memoization libraries.
